### PR TITLE
Implement GELU as function op

### PR DESCRIPTION
**Pull request by pranshupant**
### Description
These changes have been made to support the GELU operator as a function op.

### Motivation and Context
Support for [GELU: Gaussian Error Linear Unit](https://paperswithcode.com/method/gelu) activation function, which was requested in #4933. 
#4423 also mentions this under the new ops section of `Contributions Welcome`.

As per the discussion in #4933, I have added GELU as a context-dependent function-op, that uses the attribute `approximate` to return one of the two possible function-body definitions.

The first function definition is the regular GELU:
`GELU(x)=x∗Φ(x) = 0.5 * x * (1 + erf(x / sqrt(2)))`

The second is the fast approximation based on `tanh`:
`GELU(x)=0.5 ∗ x ∗ (1+Tanh( sqrt(2/π) ∗ (x + 0.044715 ∗ x^3)))`

This implementation uses the [PyTorch docs for GELU](https://pytorch.org/docs/stable/generated/torch.nn.GELU.html?highlight=gelu#torch.nn.GELU) as a reference.

PS: I also refactored `onnx/defs/math/defs.cc` to bring the operator implementation of `mish` right next to its doc string.

### CODE REVIEW

File: onnx/backend/test/case/node/gelu.py
Comment (xadupre) [inline onnx/backend/test/case/node/gelu.py:16]: I would mention tanh in the method name since it is not the default behaviour.

File: onnx/backend/test/case/node/gelu.py
Comment (gramalingam) [inline onnx/backend/test/case/node/gelu.py:21]: We need a reference implementation. We can just move this computation to the reference implementation.

Comment (gramalingam) [reply onnx/backend/test/case/node/gelu.py:21]: Eg., see the [elu implementation](https://github.com/onnx/onnx/blob/main/onnx/reference/ops/op_elu.py)

File: onnx/backend/test/case/node/gelu.py
Comment (pranshupant) [reply onnx/backend/test/case/node/gelu.py]: Done. Changed the name to export_gelu_tanh

File: onnx/backend/test/case/node/gelu.py
Comment (pranshupant) [reply onnx/backend/test/case/node/gelu.py:21]: Sorry for the delayed response, was AFK for the last 2 weeks. Added reference op implementation to onnx/reference/ops/op_gelu.py

File: onnx/reference/ops/op_gelu.py
Comment (xadupre) [inline onnx/reference/ops/op_gelu.py:1]: Since it is defined as a function, this file should not be needed. ReferenceEvaluator will inline the operator. Plus, it is not used right now because it is not imported in https://github.com/onnx/onnx/blob/main/onnx/reference/ops/_op_list.py. So removing it should not change anything.

Comment (pranshupant) [reply onnx/reference/ops/op_gelu.py:1]: @xadupre I'm a bit confused... I only added this file since @gramalingam mentioned this in his review above. Is this file not required then?

File: onnx/backend/test/case/node/gelu.py
Comment (xadupre) [reply onnx/backend/test/case/node/gelu.py:21]: It is needed when the body function is missing or if the onnx implementation is really slow. Otherwise, that should be fine since this PR https://github.com/onnx/onnx/pull/5246 was merged (yesterday).

File: onnx/reference/ops/op_gelu.py
Comment (gramalingam) [reply onnx/reference/ops/op_gelu.py:1]: Sorry about the confusion, my mistake. Please ignore my suggestion. The function-definition makes the reference implementation unnecessary. (I was thinking about the test-case generation logic, which does require an implementation. But let's leave it as is.)

File: onnx/defs/math/defs.cc
Comment (justinchuby) [inline onnx/defs/math/defs.cc:596]:
```suggestion
              XCubed = Pow ( X, ThreeCast)
```

nit?

File: onnx/defs/math/defs.cc
Comment (justinchuby) [inline onnx/defs/math/defs.cc:583]:
```suggestion
  if (approximate == "tanh") {
```

nit: I would spell it out and avoid abbreviations

File: onnx/test/automatic_upgrade_test.py
Comment (justinchuby) [inline onnx/test/automatic_upgrade_test.py:467]:
```suggestion
    def test_Gelu_approximate_tanh(self) -> None:
```

consider

File: onnx/test/automatic_upgrade_test.py
Comment (justinchuby) [inline onnx/test/automatic_upgrade_test.py:470]:
```suggestion
    def test_Gelu(self) -> None:
```

File: docs/Changelog.md
Comment (justinchuby) [inline docs/Changelog.md:23922]: Also support bfloat16?

File: docs/Operators.md
Comment (justinchuby) [inline docs/Operators.md:9418]:
```suggestion
  $y = 0.5 * x * (1 + Tanh(sqrt(2/\pi) * (x + 0.044715 * x^3)))$ is used and applied
```

optional: Possible to render math here for readability

File: docs/Operators.md
Comment (justinchuby) [inline docs/Operators.md:9430]:
```suggestion
<dd>Gelu approximation algorithm: `"tanh"`, `"none"` (default). `"none"`: do not use approximation. `"tanh"`: use tanh approximation.</dd>
```

File: onnx/defs/math/defs.cc
Comment (pranshupant) [reply onnx/defs/math/defs.cc]: Done

File: onnx/defs/math/defs.cc
Comment (pranshupant) [reply onnx/defs/math/defs.cc]: Done

File: onnx/test/automatic_upgrade_test.py
Comment (pranshupant) [reply onnx/test/automatic_upgrade_test.py]: Done

File: onnx/test/automatic_upgrade_test.py
Comment (pranshupant) [reply onnx/test/automatic_upgrade_test.py]: Done

File: docs/Changelog.md
Comment (pranshupant) [reply docs/Changelog.md]: Made these changes in onnx/defs/math/defs.cc and recompiled to update Changelog

File: docs/Operators.md
Comment (pranshupant) [reply docs/Operators.md]: Made these changes in onnx/defs/math/defs.cc and recompiled to update Changelog

File: docs/Operators.md
Comment (pranshupant) [reply docs/Operators.md]: Made these changes in onnx/defs/math/defs.cc and recompiled to update Operators.md

**Comment by pranshupant** [comment]
The Windows CI tests are failing for ORT. Based on the test logs it seems that onnxruntime doesn't have the ops registered for Opset 20 (which is expected). Is there a way to disable these tests for new ops?
Also, I feel the ORT tests should give a warning in case onnxruntime is not installed locally instead of just being skipped.

**Comment by pranshupant** [comment]
I referenced the following PR for a recent op submission #5010 and noticed a similar issue with ORT tests. Subsequently, I added `gelu` to the backend test exclude list. Please let me know if this was the right thing to do. 
All the tests pass now @gramalingam @justinchuby @xadupre

**Review by xadupre — APPROVED** [review]

**Comment by xadupre** [comment]
There are some lint issues with C++. You can use clang-format to fix them.

**Comment by justinchuby** [comment]
In _theory_ lintrunner knows which files are modified and knows to run lint on those files. Not sure what went wrong this time

**Comment by pranshupant** [comment]
@xadupre Fixed C++ linting issues. Should be ready to merge now.

**Comment by pranshupant** [comment]
@xadupre @justinchuby @gramalingam 
Let me know if I need to make any other changes before merging. Thanks!

**Comment by justinchuby** [comment]
The PR is approved. You can wait for it to be merged by maintainers (if there are no follow up comments). Thanks!

**Review by jcwchen — COMMENTED** [review]
https://github.com/onnx/onnx/pull/5390 has been merged first and so the documents (Changelog.md and Operators) need to be updated again with the latest main branch. Please regenerate these documents again by `python onnx/defs/gen_doc.py` to pass CI. Thanks!

**Review by gramalingam — APPROVED** [review]

**Review by jcwchen — APPROVED** [review]
Thank you for updating documents accordingly and now they all pass by CI check! One last thing -- please fix DCO (signoff every commit) and then this PR is ready to go.

File: .azure-pipelines/Linux-CI.yml
Comment (xadupre) [inline .azure-pipelines/Linux-CI.yml:3]: Is it related to this PR?

Comment (jcwchen) [reply .azure-pipelines/Linux-CI.yml:3]: The code here is from main branch and so updating branch in the PR should hide these irrelevant changes..

Comment (gramalingam) [reply .azure-pipelines/Linux-CI.yml]: Can we resolve this, so it can be merged? Or, does anyone have a concern? (Seems to me it should happen automatically in git)

**Comment by liqunfu** [comment]
@pranshupant, thank you for contributing to ONNX. Now I am preparing ONNX 1.15.0 release and integrating the release with ORT, I wonder if you have a plan to implement the runtime kernel for this new op in ORT? If so, I will let you know once ORT has ONNX 1.15.0 integrated. I understand that Gelu is a function op. Its implementation on ORT is optional. Thank you in advance.

**Comment by pranshupant** [comment]
> @pranshupant, thank you for contributing to ONNX. Now I am preparing ONNX 1.15.0 release and integrating the release with ORT, I wonder if you have a plan to implement the runtime kernel for this new op in ORT? If so, I will let you know once ORT has ONNX 1.15.0 integrated. I understand that Gelu is a function op. Its implementation on ORT is optional. Thank you in advance.

@liqunfu I wasn't planning to but I can consider it depending on the effort required.  Could you elaborate on that or point me to resources for making such a submission?

**Comment by liqunfu** [comment]
Thank you @pranshupant ! It generally requires add compute code for cpu and other execution providers. Because you have made Gelu a function op, runtime call execute the op without need of a compute kernel for it. I will add an optional work item for the ORT team to see it need to be implemented by the team. Thanks again! Liqun

**Pull request merged**
